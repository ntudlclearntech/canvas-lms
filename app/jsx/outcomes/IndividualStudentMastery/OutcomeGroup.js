/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import PropTypes from 'prop-types'
import I18n from 'i18n!outcomes'
import Flex, { FlexItem } from '@instructure/ui-layout/lib/components/Flex'
import View from '@instructure/ui-layout/lib/components/View'
import ToggleDetails from '@instructure/ui-toggle-details/lib/components/ToggleDetails'
import List, { ListItem } from '@instructure/ui-elements/lib/components/List'
import Pill from '@instructure/ui-elements/lib/components/Pill'
import Text from '@instructure/ui-elements/lib/components/Text'
import IconArrowOpenDown from '@instructure/ui-icons/lib/Solid/IconArrowOpenDown'
import IconArrowOpenRight from '@instructure/ui-icons/lib/Solid/IconArrowOpenRight'
import Outcome from './Outcome'
import * as shapes from './shapes'

const spacyIcon = (expanded) => () => {
  const Icon = expanded ? IconArrowOpenDown : IconArrowOpenRight
  return (
    <View padding="0 0 0 small"><Icon /></View>
  )
}

const outcomeGroupHeader = (outcomeGroup, numMastered, numGroup) => (
  <Flex justifyItems="space-between" padding="medium x-small">
    <FlexItem padding="0 x-small 0 0"><Text size="large" weight="bold">{ outcomeGroup.title }</Text></FlexItem>
    <FlexItem><Pill text={I18n.t('%{numMastered} of %{numGroup} Mastered', { numMastered, numGroup })} /></FlexItem>
  </Flex>
)

export default class OutcomeGroup extends React.Component {
  static propTypes = {
    outcomeGroup: shapes.outcomeGroupShape.isRequired,
    outcomes: PropTypes.arrayOf(shapes.outcomeShape).isRequired
  }

  constructor () {
    super()
    this.handleToggle = this.handleToggle.bind(this)
    this.state = { expanded: false }
  }

  contract () {
    this.setState({ expanded: false }, () => {
      if (this.outcomes) {
        this.outcomes.forEach((o) => {
          if (o) {
            o.contract()
          }
        })
      }
    })
  }

  expand () {
    this.setState({ expanded: true }, () => {
      if (this.outcomes) {
        this.outcomes.forEach((o) => {
          if (o) {
            o.expand()
          }
        })
      }
    })
  }

  handleToggle (_event, expanded) {
    this.setState({ expanded })
  }

  render () {
    const { outcomes, outcomeGroup } = this.props
    const numMastered = outcomes.filter((o) => o.mastered).length
    const numGroup = outcomes.length

    this.outcomes = []

    return (
      <View as="div" className="outcomeGroup" borderWidth="small" background="default" borderRadius="large">
        <ToggleDetails
          divider="solid"
          icon={spacyIcon(false)}
          iconExpanded={spacyIcon(true)}
          summary={outcomeGroupHeader(outcomeGroup, numMastered, numGroup)}
          expanded={this.state.expanded}
          onToggle={this.handleToggle}
          fluidWidth
        >
          <View as="div" borderWidth="small 0 0 0">
            <List variant="unstyled" divider="solid">
              {
                outcomes.map((outcome) => (
                  <ListItem key={outcome.id} margin="0">
                    <Outcome outcome={outcome} ref={(el) => this.outcomes.push(el)}/>
                  </ListItem>
                ))
              }
            </List>
          </View>
        </ToggleDetails>
      </View>
    )
  }
}
