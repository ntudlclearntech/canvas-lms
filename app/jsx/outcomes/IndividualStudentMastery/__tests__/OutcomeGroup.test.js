/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import { render, shallow } from 'enzyme'
import { Set } from 'immutable'
import OutcomeGroup from '../OutcomeGroup'

const defaultProps = (props = {}) => (
  Object.assign({
    outcomeGroup: {
      id: 10,
      title: 'My group'
    },
    outcomes: [
      {
        id: 1,
        expansionId: 100,
        mastered: false,
        ratings: [
          { description: 'My first rating' },
          { description: 'My second rating' }
        ],
        results: [
          {
            id: 1,
            score: 1,
            percent: 0.1,
            assignment: {
              id: 1,
              html_url: 'http://foo',
              name: 'My assignment',
              submission_types: 'online_quiz'
            }
          }
        ],
        title: 'My outcome'
      }
    ],
    expanded: false,
    expandedOutcomes: Set(),
    onExpansionChange: () => {}
  }, props)
)

it('renders the OutcomeGroup component', () => {
  const wrapper = shallow(<OutcomeGroup {...defaultProps()}/>)
  expect(wrapper.debug()).toMatchSnapshot()
})

describe('header', () => {
  it('includes the outcome group name', () => {
    const wrapper = shallow(<OutcomeGroup {...defaultProps()}/>)
    const header = wrapper.find('ToggleDetails')
    const summary = render(header.prop('summary'))
    expect(summary.text()).toMatch('My group')
  })
})

it('includes the individual outcomes', () => {
  const wrapper = shallow(<OutcomeGroup {...defaultProps()} />)
  expect(wrapper.find('Outcome')).toHaveLength(1)
})

it('renders correctly expanded', () => {
  const wrapper = shallow(<OutcomeGroup {...defaultProps()} expanded />)
  expect(wrapper.debug()).toMatchSnapshot()
})

it('passes expanded=true to the child outcome when expanded', () => {
  const props = defaultProps()
  props.expandedOutcomes = Set([100])
  const wrapper = shallow(<OutcomeGroup {...props} />)
  expect(wrapper.find('Outcome').prop('expanded')).toBe(true)
})

describe('handleToggle()', () => {
  it('calls the correct onExpansionChange callback', () => {
    const props = defaultProps()
    props.onExpansionChange = jest.fn()
    const wrapper = shallow(<OutcomeGroup {...props} />)
    wrapper.instance().handleToggle(null, true)
    expect(props.onExpansionChange).toBeCalledWith('group', 10, true)
  })
})
