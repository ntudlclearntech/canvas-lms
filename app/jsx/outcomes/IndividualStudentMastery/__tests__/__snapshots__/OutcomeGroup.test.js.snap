// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly expanded 1`] = `
"<View as=\\"div\\" className=\\"outcomeGroup\\" borderWidth=\\"small\\" background=\\"default\\" borderRadius=\\"large\\" display=\\"auto\\">
  <ToggleDetails divider=\\"solid\\" icon={[Function]} iconExpanded={[Function]} summary={{...}} expanded={true} onToggle={[Function]} fluidWidth={true} variant=\\"default\\" size=\\"medium\\" iconPosition=\\"start\\" defaultExpanded={false}>
    <View as=\\"div\\" borderWidth=\\"small 0 0 0\\" display=\\"auto\\">
      <List variant=\\"unstyled\\" divider=\\"solid\\" as=\\"ul\\" margin=\\"none\\" delimiter=\\"none\\" size=\\"medium\\">
        <ListItem margin=\\"0\\">
          <Outcome outcome={{...}} expanded={false} onExpansionChange={[Function]} />
        </ListItem>
      </List>
    </View>
  </ToggleDetails>
</View>"
`;

exports[`renders the OutcomeGroup component 1`] = `
"<View as=\\"div\\" className=\\"outcomeGroup\\" borderWidth=\\"small\\" background=\\"default\\" borderRadius=\\"large\\" display=\\"auto\\">
  <ToggleDetails divider=\\"solid\\" icon={[Function]} iconExpanded={[Function]} summary={{...}} expanded={false} onToggle={[Function]} fluidWidth={true} variant=\\"default\\" size=\\"medium\\" iconPosition=\\"start\\" defaultExpanded={false}>
    <View as=\\"div\\" borderWidth=\\"small 0 0 0\\" display=\\"auto\\">
      <List variant=\\"unstyled\\" divider=\\"solid\\" as=\\"ul\\" margin=\\"none\\" delimiter=\\"none\\" size=\\"medium\\">
        <ListItem margin=\\"0\\">
          <Outcome outcome={{...}} expanded={false} onExpansionChange={[Function]} />
        </ListItem>
      </List>
    </View>
  </ToggleDetails>
</View>"
`;
